### DISCUSSION TITLE
Feature gap?

### DISCUSSION THREAD

**Discussion by almazik** [ory/kratos#941] [Q&A] [answered]
We are looking for a new identity management solution to replace our home-grown one. I really liked the simplicity of Kratos, but it looks that the feature gap is still substantial. We are quite a large business with millions of user identities.

We will consider hiring developers to support OSS development of the Ory ecosystem if we commit to this solution.

1. Are there plans to provide hooks on registration/login/logouts/identity updates? We have a Kafka-centric event-driven microservices solution, and it is essential for us to publish the information to Kafka whenever the observable state of the system changes. It's not a blocker as we can set up a Kafka Connect on top of the underlying DB, but that makes it dependent on the internals of the Kratos.
2. Some of our identity traits fields may be the subject to identity validation (know your customer, KYC). Once they are validated we do not allow the identity to be edited by the end-user. Is there anything in the roadmap to support this scenario?
3. As the extension of the previous point - the ability to freeze certain identity fields after registration.
4. Do you have plans to support device authentication tokens - the token associated with the app installed on the iOS/Android device and used to initialize session? Additional functionality - the ability for user/admin to list all authenticated devices, revoke authentication token(s).
5. Do have plans to support user migration scenarios similar to [auth0 automatic migrations](https://auth0.com/docs/users/configure-automatic-migration-from-your-database)? E.g. configure a callback to the legacy system when you cannot find the corresponding user, and store the identity on successful legacy system response.
6. Do you have protection (or plans) for brute force attacks (repetitive login attempts)? Leaked database scans?

**Comment by aeneasr** [comment, marked as answer]
Thank you for your questions and offer of contributions. Almost all of your questions can be answered by taking a quick glance at the GitHub issue tracking / GitHub search, or by searching in the docs. By doing some of the research yourself, you can help keep the noise for maintainers low, helping them focus on coding tasks. We appreciate your participation nonetheless.

> Are there plans to provide hooks on registration/login/logouts/identity updates?

https://github.com/ory/kratos/issues?q=is%3Aissue+is%3Aopen+hook+rpc

> Some of our identity traits fields may be the subject to identity validation (know your customer, KYC). 

https://github.com/ory/kratos/issues/47

> As the extension of the previous point - the ability to freeze certain identity fields after registration.

https://github.com/ory/kratos/issues/47

> Do you have plans to support device authentication tokens

Do you mean MFA/2FA/FIDO2?

>  Additional functionality - the ability for user/admin to list all authenticated devices, revoke authentication token(s).

https://github.com/ory/kratos/issues/655

> Do have plans to support user migration scenarios similar to auth0 automatic migrations? E.g. configure a callback to the legacy system when you cannot find the corresponding user, and store the identity on successful legacy system response.

No plans yet but migration is an important piece of the puzzle. We will probably focus on importing existing users first though. An alternative to callback and custom code is probably fronting the legacy system with ORY Hydra (OAuth2/OIDC Server) and then using that as an upstream in ORY Kratos.

> Do you have protection (or plans) for brute force attacks (repetitive login attempts)? Leaked database scans?

https://www.ory.sh/kratos/docs/concepts/security#passwords
